### PR TITLE
test: add BDD integration wave 2 (63 tests)

### PR DESCRIPTION
**Pull request by EffortlessSteven**
## BDD Integration Wave 2

Adds 63 BDD-style integration tests across 10 modules validating end-to-end behavior of the inference pipeline.

### Test Categories

| Module | Tests | Coverage |
|--------|-------|----------|
| Kernel dispatch | 8 | CPU selection, fallback, caching, trait objects |
| Configuration validation | 7 | SessionConfig fields, backends, error messages |
| Memory safety | 9 | Empty/zero inputs for quant, embedding, norm, RoPE |
| Device selection | 7 | CPU/GPU predicates, probe, graceful fallback |
| Error propagation | 6 | BitNetError conversions, descriptive messages |
| Engine state machine | 5 | Valid/invalid state transitions |
| Sampling pipeline | 4 | Greedy, seeded, repetition penalty |
| Stop criteria | 5 | Token IDs, EOS, max tokens, strings |
| KV cache validation | 4 | Config validation, initial state |
| Logits + session + concurrency | 8 | Temperature, softmax, top-k, uniqueness |

### Structure

All tests follow Given/When/Then BDD pattern:
```rust
#[test]
fn given_invalid_config_when_creating_engine_then_returns_error() {
    // Given: An invalid inference configuration
    // When: We try to create an engine  
    // Then: It returns a descriptive error
}
```

### Verification

- `cargo test -p bitnet-tests --test bdd_integration_wave2 --no-default-features --features cpu` — 63/63 pass
- `cargo fmt` — clean
- `cargo clippy -D warnings` — clean

### CODE REVIEW

**Comment by EffortlessSteven** [comment]
Closing: stale test wave  significant drift from main, would need full rewrite to rebase.
